### ISSUE TITLE
messageDeleted undefined for messages before bot start/reboot?

### ISSUE THREAD

**Issue by DJScias** [hydrabolt/discord.js#484]
Hello there,

I've ran a few tests on a local solution of a bot and have used the following:
````
testbot.on("messageDeleted", function(msg) {
    if (msg.server) {
        console.log(msg);
    }
});
````

All the messages that are made after the bot joins and are then deleted return the obvious Message object.

However if you try to delete a msg that was made before the bot joins / has rebooted (something which I am indeed trying to do) returns with the following:
````
if (msg.server) {
       ^

TypeError: Cannot read property 'server' of null
`````

Is there a workaround to make "messageDeleted" work on pre-reboot messages or am I doing something wrong here to call the correct object?

Any information would be greatly appreciated.

Kind regards,
DJScias

**Comment by acdenisSK** [comment]
Yes and you have two (or more ¯\_(ツ)_/¯) solutions: check if `msg` isn't `null` or suround it around a try/catch

**Comment by DJScias** [comment]
Hello there @acdenisSK ,

Correct me if I'm wrong.. But wouldn't do both those solutions just do checks on "if msg exists" and in the end still not give me any information in regards to the Message object?

Considering I'm trying to obtain the [Message object's id](http://discordjs.readthedocs.io/en/latest/docs_channel.html#id) on the deletion of said message. Whether or not it was created before or after the bot's start/reboot.

Kind regards,
DJScias

**Comment by acdenisSK** [comment]
Yeah i just realised the other part. Well there isn't a solution for that except having your bot be 24/7 on and the reason is d.js needs to get the information first from some source, and that source would be the WS (WebSocket) which will aid d.js for getting it but because you're trying to get info about a entity that was sent already from the WS before the bot got the oppurtunity to boot up. And those two can be fail-switches so node won't complain about a non exisiting object.

**Comment by DJScias** [comment]
I understand.

I thought there would have been some sort of workaround to like check the last 10 messages with "getChannelLogs" and add their Message objects to the cache manually or something.

I know the above sounds like "But that would only make you able to have 10 previous message in the cache", but the work-idea of the bot is just to check a channel which never has more than 10 messages, ever.

Kind regards,
DJScias

**Comment by acdenisSK** [comment]
`getChannelsLogs` uses the REST api which just gives a xx amount of messages but they aren't put into the cache but you can modify d.js's messages cache (`client.messages`) with the [add](http://discordjs.readthedocs.io/en/indev/docs_cache.html#add-data) method but i don't know if this solution could break something but i guess if it's not being "24/7 on" then try it.

**Comment by DJScias** [comment]
Well, the thing is that it is indeed 24/7 on as it's being ran by a Linux server (DigitalOcean).

However, sometimes I do updates to the bot and then I have to restart the process. Now, every time I restart the process (aka reboot) the messages before that reboot time, but which were made after the bot joining the server, also return undefined.

So yes, that is the issue / dilemma that is presented here with myself.

Kind regards,
DJScias

**Comment by macdja38** [comment]
So, I believe you are trying to say when messages that were sent before the last time the bot was online are deleted you don't get the message. And that messages that are sent before the ready event don't have a server property. Those are two quirks of how things kinda have to happen.

1) Message from before bot enable or old message is deleted.
Discord does not send the contents of a deleted message, just it's id. so if the bot does not have the message cached it can't tell what the message used to contain. If you want to get around this you are going to need to make your own cach (a database) and potentially modify  the lib so it emits an id when no message is present.

2) Server object is not attatched to messages deleted before read.
Well... before the ready event the bot has not recieved all the servers... so the bot does not have that server cached so it can't be attached. You could try a system where you only kill the old bot once the new bot is ready.

**Comment by DJScias** [comment]
@macdja38 Hello there.

Sorry it's taken me a few days before I replied.

You mention in 1) that in the case of an old/before bot message getting deleted it does not send the contents, but it does send the id.

I've tried this with `testbot.on("messageDeleted", function(msg) { }` but it's just sending nothing.. Or is there a different syntax to get the deleted message id on `messageDeleted`? As I honestly don't need the contents of the deleted message but only the message id. So if what you say is true, then that'd fix my issues.

Kind regards,
DJSCias

**Comment by GexoXYZ** [comment]
You could always listen for the raw payloads

**Comment by DJScias** [comment]
Hello there @GexoXYZ 

Any possible documentation pointing me in the right direction in regards to that?

Kind regards,
DJScias

**Comment by GexoXYZ** [comment]
```js
testbot.on("raw", packet => {
    if(packet.t === "MESSAGE_DELETE") {
        let messageID = packet.d.id;
        let channelID = packet.d.channel_id;
    }
});
```
Something similar to this. (Code might not be correct, just wrote this off the top of my head)

**Comment by hydrabolt** [comment]
Fixed in [9.0.2](https://github.com/hydrabolt/discord.js/releases/tag/9.0.2)
